### ISSUE TITLE
validator: Module not found 'avj-keywords'

### ISSUE THREAD

**Issue by willfarrell** [middyjs/middy#560]
When using `serverless` with `serverless-bundle`.

> ## Warning
> If you use serverless with serverless-bundle you will run into a build issue (Module not found 'ajv-keywords') with the default configuration. See #560 for details. You can use v1.2.0 or set ajvPlugins to {} to remove all plugins and work around this issue.

https://github.com/middyjs/middy/pull/552#issuecomment-684126761

**Comment by jverce** [comment]
Having the same issue here. Webpack doesn't seem to get along well with dynamic `require`'s.
Looking into how to fix this without putting the burden on users.

**Comment by willfarrell** [comment]
I'm open to solutions, however, it must be able to run in node v10.

**Comment by willfarrell** [comment]
v1.3.2 has the fix in it. cc @renanwilliam

**Comment by renanwilliam** [comment]
Great, updated our project and everything back to work like a charm :)

**Comment by ThomasKiec** [comment]
@willfarrell  I have the same issue but in my case this happens since the fixes in v1.3.2. Before v1.3.2 everything worked just fine

**Comment by willfarrell** [comment]
@ThomasKiec Can you describe your build process, using any extra serverless plugins that might touch the code?

**Comment by ThomasKiec** [comment]
In my current project I use serverless with the  serverless-webpack plugin.

**Comment by renanwilliam** [comment]
No problems here but I'm using ^1.4.0

**Comment by anthilton** [comment]
I maybe having the same issues as @ThomasKiec here. 
Setup = monorepo, typescript, serverless.

When using validator version 1.4.0 with webpack 4 and serverless-webpack I was getting a whole bunch of WARNINGS, but it still bundled and worked. 
An example warning:
WARNING in ../../node_modules/ajv-errors/README.md 1:0
Module parse failed: Unexpected character '#' (1:0)

Then I tried to upgrade from webpack 4 to webpack 5 and am now getting 
ERROR in ../../node_modules/ajv-errors/LICENSE 1:4
Module parse failed: Unexpected token (1:4)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file.

But as it is an error it won't bundle.

I seem to have gotten around this for now with the following ignore loader regex, though it is hacky
{ test: /(?=(ajv-.*))((?=.*(\.(?!(js|json)\b).*))|(?=.*(LICENSE)))/, loader: 'ignore-loader' },

1 thing I did notice was that if you update following it works .. (well it did for me and makes no sense why)
onst pckJson = require(`../../ajv-${p}/package.json`)
pluginsInstances[p] = require(`../../ajv-${p}/${pckJson.main}`)
	  
to:

const pckJson = require(`../../ajv-${p}/package.json`);
const mainFile = `../../ajv-${p}/${pckJson.main}`;
pluginsInstances[p] = require(mainFile);

But I cannot change the node_module on my CI server.

Hopefully this will help someone else :)

**Comment by yhaiovyi** [comment]
Same issue, ton of output with `Module parse failed: Unexpected token`

**Comment by gamercraftlab** [comment]
Same issue on our end

**Comment by shanetierney** [comment]
> I maybe having the same issues as @ThomasKiec here.
> Setup = monorepo, typescript, serverless.
> 
> When using validator version 1.4.0 with webpack 4 and serverless-webpack I was getting a whole bunch of WARNINGS, but it still bundled and worked.
> An example warning:
> WARNING in ../../node_modules/ajv-errors/README.md 1:0
> Module parse failed: Unexpected character '#' (1:0)
> 
> Then I tried to upgrade from webpack 4 to webpack 5 and am now getting
> ERROR in ../../node_modules/ajv-errors/LICENSE 1:4
> Module parse failed: Unexpected token (1:4)
> You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file.
> 
> But as it is an error it won't bundle.
> 
> I seem to have gotten around this for now with the following ignore loader regex, though it is hacky
> `{ test: /(?=(ajv-.*))((?=.*(\.(?!(js|json)\b).*))|(?=.*(LICENSE)))/, loader: 'ignore-loader' },`
> 
> 1 thing I did notice was that if you update following it works .. (well it did for me and makes no sense why)
> `const pckJson = require(`../../ajv-${p}/package.json`) pluginsInstances[p] = require(`../../ajv-${p}/${pckJson.main}`)`
> 
> to:
> 
> `const pckJson = require(`../../ajv-${p}/package.json`); const mainFile = `../../ajv-${p}/${pckJson.main}`; pluginsInstances[p] = require(mainFile);`
> 
> But I cannot change the node_module on my CI server.
> 
> Hopefully this will help someone else :)

Having this same issue

**Comment by willfarrell** [comment]
Going to re-open for now to ensure this is well tested in the next version. I may reach out for help testing.

**Comment by jarrodldavis** [comment]
Using the excellent `[patch-package](https://www.npmjs.com/package/patch-package)`, I was able to fix this issue using webpack's `require.context` API:

```patch
diff --git a/node_modules/@middy/validator/index.js b/node_modules/@middy/validator/index.js
index 36abf97..e001846 100644
--- a/node_modules/@middy/validator/index.js
+++ b/node_modules/@middy/validator/index.js
@@ -138,7 +138,9 @@ function getPlugins (p) {
   try {
     /* Fixes #560: Webpack needs explicit paths for dynamic imports */
     const pckJson = require(`../../ajv-${p}/package.json`)
-    pluginsInstances[p] = require(`../../ajv-${p}/${pckJson.main}`)
+    const mainFile = `../../ajv-${p}/${pckJson.main}`
+    const pluginRequire = require.context('../../', true, /ajv-\w+\/.+\.js$/)
+    pluginsInstances[p] = pluginRequire(mainFile)
   } catch (e) {
     console.error(`Error getting plugins ${e.message}`)
   }
```

After setting up `patch-package` in your project, you'll need to replicate the above changes in `node_modules/@middy/validator/index.js` and run `npx patch-package @middy/validator` to fix your copy of `@middy/validator`.

If, like me, you don't want `ajv-i18n` included in your Lambda bundle, you can update `node_modules/@middy/validator/index.js` with these changes:

```patch
diff --git a/node_modules/@middy/validator/index.js b/node_modules/@middy/validator/index.js
index 36abf97..ab19f06 100644
--- a/node_modules/@middy/validator/index.js
+++ b/node_modules/@middy/validator/index.js
@@ -138,7 +138,9 @@ function getPlugins (p) {
   try {
     /* Fixes #560: Webpack needs explicit paths for dynamic imports */
     const pckJson = require(`../../ajv-${p}/package.json`)
-    pluginsInstances[p] = require(`../../ajv-${p}/${pckJson.main}`)
+    const mainFile = `../../ajv-${p}/${pckJson.main}`
+    const pluginRequire = require.context('../../', true, /ajv-(?!i18n)\w+\/.+\.js$/)
+    pluginsInstances[p] = pluginRequire(mainFile)
   } catch (e) {
     console.error(`Error getting plugins ${e.message}`)
   }
```

If you'd like to exclude other Ajv plugins, update the regular expression accordingly:

```patch
diff --git a/node_modules/@middy/validator/index.js b/node_modules/@middy/validator/index.js
index 36abf97..647d5a6 100644
--- a/node_modules/@middy/validator/index.js
+++ b/node_modules/@middy/validator/index.js
@@ -138,7 +138,9 @@ function getPlugins (p) {
   try {
     /* Fixes #560: Webpack needs explicit paths for dynamic imports */
     const pckJson = require(`../../ajv-${p}/package.json`)
-    pluginsInstances[p] = require(`../../ajv-${p}/${pckJson.main}`)
+    const mainFile = `../../ajv-${p}/${pckJson.main}`
+    const pluginRequire = require.context('../../', true, /ajv-(?!i18n|keywords)\w+\/.+\.js$/)
+    pluginsInstances[p] = pluginRequire(mainFile)
   } catch (e) {
     console.error(`Error getting plugins ${e.message}`)
   }
```

**Comment by willfarrell** [comment]
Confirmed, this is a non-issue in v2
